channels/audin/server/audin.c

```diff
@@ -21,356 +22,206 @@
- *
- * @return 0 on success, otherwise a Win32 error code
- */
-static UINT audin_server_send_open(audin_server* audin, wStream* s)
+static UINT audin_server_recv_data_incoming(audin_server_context* context, wStream* s,
+                                            const SNDIN_PDU* header)
 {
-	ULONG written;
+	SNDIN_DATA_INCOMING pdu = { 0 };
+	UINT error = CHANNEL_RC_OK;
 
-	WINPR_ASSERT(audin);
-	if (audin->context.selected_client_format < 0)
-	{
-		WLog_ERR(TAG, "audin->context.selected_client_format = %" PRIdz,
-		         audin->context.selected_client_format);
-		return ERROR_INVALID_DATA;
-	}
+	WINPR_ASSERT(context);
+	WINPR_ASSERT(header);
+
+	pdu.Header = *header;
+
+	IFCALLRET(context->IncomingData, error, context, &pdu);
+	if (error)
+		WLog_ERR(TAG, "context->IncomingData failed with error %" PRIu32 "", error);
 
-	audin->opened = TRUE;
-	Stream_SetPosition(s, 0);
-	Stream_Write_UINT8(s, MSG_SNDIN_OPEN);
-	Stream_Write_UINT32(s, audin->context.frames_per_packet); /* FramesPerPacket (4 bytes) */
-	WINPR_ASSERT(audin->context.selected_client_format >= 0);
-	WINPR_ASSERT(audin->context.selected_client_format <= UINT32_MAX);
-	Stream_Write_UINT32(
-	    s, (UINT32)audin->context.selected_client_format); /* initialFormat (4 bytes) */
-	/*
-	 * [MS-RDPEAI] 3.2.5.1.6
-	 * The second format specify the format that SHOULD be used to capture data from
-	 * the actual audio input device.
-	 */
-	Stream_Write_UINT16(s, 1);             /* wFormatTag = PCM */
-	Stream_Write_UINT16(s, 2);             /* nChannels */
-	Stream_Write_UINT32(s, 44100);         /* nSamplesPerSec */
-	Stream_Write_UINT32(s, 44100 * 2 * 2); /* nAvgBytesPerSec */
-	Stream_Write_UINT16(s, 4);             /* nBlockAlign */
-	Stream_Write_UINT16(s, 16);            /* wBitsPerSample */
-	Stream_Write_UINT16(s, 0);             /* cbSize */
-	WINPR_ASSERT(Stream_GetPosition(s) <= ULONG_MAX);
-	return WTSVirtualChannelWrite(audin->audin_channel, (PCHAR)Stream_Buffer(s),
-	                              (ULONG)Stream_GetPosition(s), &written)
-	           ? CHANNEL_RC_OK
-	           : ERROR_INTERNAL_ERROR;
+	return error;
 }
 
-/**
- * Function description
- *
- * @return 0 on success, otherwise a Win32 error code
- */
-static UINT audin_server_recv_open_reply(audin_server* audin, wStream* s, UINT32 length)
+static UINT audin_server_recv_data(audin_server_context* context, wStream* s,
+                                   const SNDIN_PDU* header)
 {
-	UINT32 Result;
-	UINT success = CHANNEL_RC_OK;
+	SNDIN_DATA pdu = { 0 };
+	UINT error = CHANNEL_RC_OK;
 
-	WINPR_ASSERT(audin);
-	if (length < 4)
-	{
-		WLog_ERR(TAG, "error parsing version info: expected at least 4 bytes, got %" PRIu32 "",
-		         length);
-		return ERROR_INVALID_DATA;
-	}
+	WINPR_ASSERT(context);
+	WINPR_ASSERT(header);
+
+	pdu.Header = *header;
 
-	Stream_Read_UINT32(s, Result);
-	IFCALLRET(audin->context.OpenResult, success, &audin->context, Result);
+	pdu.DataSize = Stream_GetRemainingLength(s);
```

Comment:
Might be a good idea to usw a `wStream*` in the `SNDIN_DATA` pdu and initialize it as a static stream?

Reply:
Did that in the last push.